### PR TITLE
[Form] set unique block prefix with any namespace

### PR DESCRIPTION
**Pull request by andrey1s**
| Q | A |
| --- | --- |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | - |
| License | MIT |
| Doc PR | - |

if there is no block `hidden_locale_row` when error

```
An exception has been thrown during the rendering of a template ("Unable to
render the form as none of the following blocks exist: "_form_texts_0_row",
"hidden_locale_row", "hidden_locale_row", "form_row", "form_row".") in
form_div_layout.html.twig at line 312.
```

### CODE REVIEW

**Comment by enumag** [comment]
:+1: I've just noticed similar error in my own application. The problem seems to occur when a prefix is repeated in the block_prefixes array. The exception is thrown too early - in the example from @andrey1s the block form_row certainly exists even though the exception says otherwise.

**Comment by webmozart** [comment]
Hi, thanks for reporting this problem :) How can it happen that a block name is repeated in the chain? Can you provide code that reproduces your problem?

**Comment by andrey1s** [comment]
for example
```php
//src/AppBundle/Controller/IndexController.php
<?php

namespace AppBundle\Controller;

use Sensio\Bundle\FrameworkExtraBundle\Configuration\Route;
use Symfony\Bundle\FrameworkBundle\Controller\Controller;
use Symfony\Component\HttpFoundation\Request;
use AppBundle\Form\Type\Other\SimpleType;

class IndexController extends Controller
{
    /**
     * @Route("/", name="homepage")
     */
    public function homepageAction(Request $request)
    {
        $form = $this
            ->createFormBuilder()
            ->add('simple', SimpleType::class)
            ->getForm()
        ;

        return $this->render('@App/Index/homepage.html.twig', ['form' => $form->createView()]);
    }
}
```

forms

```php
<?php
//src/AppBundle/Form/Type/Other/SimpleType.php
namespace AppBundle\Form\Type\Other;

use Symfony\Component\Form\AbstractType;
use AppBundle\Form\Type\SimpleType as BaseType;

class SimpleType extends AbstractType
{
    public function getParent()
    {
        return BaseType::class;
    }
}
```

```php
<?php
//src/AppBundle/Form/Type/SimpleType.php
namespace AppBundle\Form\Type;

use Symfony\Component\Form\AbstractType;
use Symfony\Component\Form\Extension\Core\Type\TextType;

class SimpleType extends AbstractType
{
    public function getParent()
    {
        return TextType::class;
    }
}
```
when there are two blocks with the same name is an error. 
We can quickly fix the error in the form for example:
```php
<?php
//src/AppBundle/Form/Type/Other/SimpleType.php
namespace AppBundle\Form\Type\Other;

use Symfony\Component\Form\AbstractType;
use AppBundle\Form\Type\SimpleType as BaseType;

class SimpleType extends AbstractType
{
    public function getParent()
    {
        return BaseType::class;
    }

    public function getBlockPrefix()
    {
        return 'text';
    }
}
```

File: src/Symfony/Component/Form/AbstractRendererEngine.php [hunk 1]
```diff
@@ -171,7 +171,7 @@ private function loadResourceForBlockNameHierarchy($cacheKey, FormView $view, ar
             // The next two if statements contain slightly duplicated code. This is by intention
             // and tries to avoid execution of unnecessary checks in order to increase performance.
 
-            if (isset($this->resources[$cacheKey][$parentBlockName])) {
+            if (isset($this->resources[$cacheKey][$parentBlockName]) && $this->resources[$cacheKey][$parentBlockName] !== false) {
```
Comment (webmozart) [inline src/Symfony/Component/Form/AbstractRendererEngine.php]: This change is missing a corresponding test.

Also you are reducing the impact of a performance optimization here. Could you benchmark the result for big forms?

**Comment by webmozart** [comment]
Thanks for the example, I understand your use case now. This conveys a bigger problem IMO.

In a template, a block (e.g. `checkbox_widget`) expects a specific block when rendering the block for the parent type (`form_widget`). By returning the block of *another* type from `getBlockPrefix()` we mess up that inheritance chain:

* Twig blocks can't expect deterministic parent blocks anymore.
* We can't short-circuit the loading of parent blocks anymore [as we did so far](https://github.com/symfony/symfony/pull/17874#discussion-diff-55326441R174), reducing the potential for performance improvements.

We should consider to recommend returning unique values from `getBlockPrefix()`.

File: src/Symfony/Component/Form/AbstractRendererEngine.php (hunk 1 above)
Comment (andrey1s) [inline src/Symfony/Component/Form/AbstractRendererEngine.php]: i can't find any test for the `Symfony\Component\Form\AbstractRendererEngine` or `Symfony\Component\Form\Extension\Templating\TemplatingRendererEngine`, could you show where are they located?

could you show the current benchmark?

**Comment by enumag** [comment]
@webmozart In that case we should modify the exception - if there is a duplicate in the array throw an exception about that instead of what is thrown currently. Probably only when a problem like this occurs to avoid unnecessary checking for non-unique prefixes for every field.

**Comment by andrey1s** [comment]
@webmozart if we need to use a unique template, then Yes, otherwise it's not necessary.
if use unique block_prefix, may be better to use by default, e.g. like this `AppBundle\Form\Type\Other\SimpleType` => `other_simple` and `AppBundle\Form\Type\SimpleType` => `simple` or `AppBundle\Form\Type\Other\Another\SimpleType` => `other_ another_simple`

or `AbstractType::getBlockPrefix()` should return an empty block name. And add only if not empty.

**Comment by webmozart** [comment]
@enumag Throwing an exception sounds like the best solution to me.

@andrey1s Changing the default block name is not possible without breaking BC.

File: src/Symfony/Component/Form/AbstractRendererEngine.php (hunk 1 above)
Comment (issei-m) [inline src/Symfony/Component/Form/AbstractRendererEngine.php]: Prefer to the yoda style like:

```php
false !== $this->resources[$cacheKey][$parentBlockName]
```

**Comment by fabpot** [comment]
IIUC, this change should not be merged but a new exception  should be added, is that correct? How do we moved forward?

**Comment by enumag** [comment]
@fabpot I think adding an exeption like [this](https://github.com/symfony/symfony/commit/d3308193beb303ac91f140a63fe772e295a231d5) solves the problem (at least the exception message is not so confusing anymore). However I was unable to write a test for it since FormRenderer class doesn't have any tests.

**Comment by fabpot** [comment]
@enumag Can you update this PR with the commit you mentioned above? Tests can be added later or by someone else if needed.

**Comment by enumag** [comment]
@fabpot I'm not the author of this PR so I'll have to create a new one instead. Wait a moment...

**Comment by xabbuh** [comment]
Does that mean we should close here in favour of #19127?

**Comment by fabpot** [comment]
Closing in favor of #19127

**Pull request closed without merging**